### DISCUSSION TITLE
How to error messages from react-select multi with RHF?

### DISCUSSION THREAD

**Discussion by wick3ds0ul** [react-hook-form/react-hook-form#8243] [Q&A] [answered]
I am using RHF with react-select and yup.

Here is the yup schema for this `field`:
```JS

import * as yup from 'yup';

export const mySchema = yup
  .object({
    title: yup.string().required('Title is required'),
    alias: yup.string().required('Alias is required'),
    author: yup.string().required('Author is required'),
    artist: yup.string().required('Artist is required'),
    description: yup.string().required('Description is required'),
    genre: yup
      .array()
      .min(2, 'Pick at least 2 genres')
      .of(
        yup.object({
          label: yup.string(),
          value: yup.string(),
          color: yup.string(),
        })
      ),
    status: yup
      .object({
        value: yup.string(),
        label: yup.string(),
        color: yup.string(),
      })
      .required('Status is required'),
    type: yup
      .object({
        value: yup.string(),
        label: yup.string(),
        color: yup.string(),
      })
      .required('Type is required'),
    year: yup
      .object({
        value: yup.string(),
        label: yup.string(),
        color: yup.string(),
      })
      .nullable()
      .required('Year is required'),
  })
  .required();


```

Here is shape of `DataOption`:
```JS 
interface DataOption {
  readonly value: string;
  readonly label: string;
  readonly color: string;
}
```
Here are example of options:

```JS
 const colourOptions: readonly DataOption[] = [
  { value: 'A', label: 'A', color: '#FF8B00' },
  { value: 'B', label: 'B', color: '#FFC400' },
  { value: 'C', label: 'C', color: '#36B37E' },
  { value: 'D', label: 'D', color: '#00875A' },
];
```

Usage of `useForm`:
```JS
  const {
    register,
    handleSubmit,
    control,
    formState: { errors },
  } = useForm<FormData>({
    resolver: yupResolver(mySchema),
  });
```
Here is `FormData`:
```JS
export type FormData = {
  title: string;
  alias: string;
  author: string;
  artist: string;
  description: string;
  genre: DataOption[];
  status: DataOption;
  type: DataOption;
  year: DataOption;
};
```

The `defaultValue` is an array.

```JS
            <Controller
              name="genre"
              control={control}
              defaultValue={[colourOptions[0], colourOptions[1]]}
              render={({ field }) => (
                <ColorSelect
                  label="Select Genre"
                  className="col-span-2"
                  field={field}
                  errorMessage={errors.genre?.values?.message}

                />
              )}
            />
```
<img width="1441" alt="Screenshot 2022-04-25 at 10 24 19 AM" src="https://user-images.githubusercontent.com/48680692/165023331-050076b8-4f0d-4cbc-98eb-ea3b3323aac9.png">

`errorMessage={errors.field?.values?.message}` works for SingleSelect.

I am getting a ts warning for error message.

**Comment by bluebill1049** [comment]
Are you using this type? https://react-hook-form.com/ts#NestedValue

**Reply by wick3ds0ul** [reply]
I have added `NestedValue` to form values. 
So now 

```JS
export type FormData = {
  title: string;
  alias: string;
  author: string;
  artist: string;
  description: string;
  genre: NestedValue<DataOption[]>;
  status: DataOption;
  type: DataOption;
  year: DataOption;
};

```
The error warning goes away from the field. But there is this warning on `useForm` and `handleSubmit`.

<img width="1114" alt="Screenshot 2022-04-26 at 2 05 48 AM" src="https://user-images.githubusercontent.com/48680692/165170533-4b115818-1bbc-467f-a8fe-98f1ad2141e4.png">
<img width="1414" alt="Screenshot 2022-04-26 at 2 06 04 AM" src="https://user-images.githubusercontent.com/48680692/165170564-631e4965-d8bc-4e4c-a5bb-d27af5186460.png">

**Comment by Moshyfawn** [comment]
Just a guess, but depending on how you generate type of your form 
```ts
useForm<FormValues>()
```
There's an issue with using 
```ts
type FormValues = yup.infer<typeof schema>
```
, if there's a `.shape` operator used.

The workaround would be to omit the `.shape` operator. There was a discussions / issue on the case recently somewhere in this repo.

Once again, I'm not sure if it's related or not

**Reply by wick3ds0ul** [reply]
I have updated the question with `FormData` and usage of `useForm` and removed `.shape` operator.

**Comment by Moshyfawn** [comment, marked as answer]
Wait, @wick3ds0ul, so your genre field is an `array` of `object`s, right? Why are your trying to access the `array.values` method and use it as an object of nested values?

As genre is an array, the correct way to access the field's error message is
```js
errors.genre?.[itemIndex]?.value?.message
```

If you're trying to access the root genre error message, it would look like
```js
errors.genre?.message
```
But RHF doesn't support the root field array error message type (TS only issue). You can access your field array error message as shown above
```js
errors.genre?.message
```
but TS will complain

ATM. the only workaround is to
```ts
@ts-expect-error insufficient rhf types
```

Expect it fixed in [the next major RHF v8](https://github.com/react-hook-form/react-hook-form/discussions/7433)

**Reply by wick3ds0ul** [reply]
I was trying to get root error message, i.e `Pick at least 2.`. After adding `@ts-expect-ignore` the TS complain is not there.

**Reply by Moshyfawn** [reply]
Keep in mind, that `@ts-expect-ignore` only shuts the TS compiler up. The type error is still there.

As I've already mentioned, we're working to bring more type-safety in RHF v8 ;)
